dataset/test/buckets_test.cpp

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+#include <gtest/gtest.h>
+
+#include "scipp/dataset/bucket.h"
+#include "scipp/dataset/dataset.h"
+#include "scipp/dataset/shape.h"
+#include "scipp/variable/bucket_model.h"
+#include "scipp/variable/operations.h"
+
+using namespace scipp;
+using namespace scipp::dataset;
+
+class DataArrayBucketTest : public ::testing::Test {
+protected:
+  using Model = variable::DataModel<bucket<DataArray>>;
+  Dimensions dims{Dim::Y, 2};
+  Variable indices = makeVariable<std::pair<scipp::index, scipp::index>>(
+      dims, Values{std::pair{0, 2}, std::pair{2, 4}});
+  Variable data =
+      makeVariable<double>(Dims{Dim::X}, Shape{4}, Values{1, 2, 3, 4});
+  DataArray buffer = DataArray(data, {{Dim::X, data + data}});
+  Variable var{std::make_unique<Model>(indices, Dim::X, buffer)};
+};
+
+TEST_F(DataArrayBucketTest, concatenate) {
+  const auto result = buckets::concatenate(var, var * (3.0 * units::one));
+  Variable out_indices = makeVariable<std::pair<scipp::index, scipp::index>>(
+      dims, Values{std::pair{0, 4}, std::pair{4, 8}});
+  Variable out_data = makeVariable<double>(Dims{Dim::X}, Shape{8},
+                                           Values{1, 2, 3, 6, 3, 4, 9, 12});
+  Variable out_x = makeVariable<double>(Dims{Dim::X}, Shape{8},
+                                        Values{2, 4, 2, 4, 6, 8, 6, 8});
+  DataArray out_buffer = DataArray(out_data, {{Dim::X, out_x}});
+  EXPECT_EQ(result,
+            Variable(std::make_unique<Model>(out_indices, Dim::X, out_buffer)));
+
+  // "in-place" append gives same as concatenate
+  buckets::append(var, var * (3.0 * units::one));
```

Comment:
Regarding the syntax, while i expect `concatenate` to be a free function, i would have assumed `append` to be a method (maybe that's because i've done too much python?).
e.g. `var.append(other_var)`

In addition, for the python bindings, i think it may be more user-friendly if users did not have to remember whether their data was `buckets` or not.
They should be able to do `sc.concatenate(var1, var2)` instead of `sc.buckets.concatenate(...)`.

Reply:
But looking at other operations that are being added, there will be `sc.buckets.scale`to scale, e.g., event data by a histogram. Without the `buckets` in the name it is unclear what it does. `sc.concatenate` would also have double meaning (did the user forget the `dim` arg, or did the intend to concatenate buckets)?

Reply:
i'm just saying that event data, and hence buckets, will be the main data format used in `scipp` at ESS. And having to prepend most functions with `buckets` feels like we are moving away from something that resembles `numpy`.

Reply:
I don't think there is another way. Whether it is `sc.buckets.concatenate` or `sc.concatenate_elements` is another question, but I think we have to express this *somehow*, since otherwise there are confusing ambiguities with the dense operations.

Reply:
Could you explain a little more why doing `scale` is unclear?
I would have expected that if I want to scale event data by a histogram, i would simply do
```
event_data / histogrammed_data
```
You are saying we cannot do this anymore?

Reply:
The element of bucketed data are, e.g., `DataArrayView`. These can be divided or added, so I we write

```python
event_data + event_data
event_data / histogram
```
The first would actually add each element (`DataArrayView`) and fail from coord mismatch, whereas we need to *concatenate* the elements to get what we actually want.

In the second case, `event_data / histogram` would usually just operate on the elements, i.e., divide each `DataArrayView` by a scalar. If the binning/bucketing is mismatching it would fail from coord/dim mismatch. `buckets::scale` would *descent* into the buckets, looking up the coords of each event in each `DataArrayView` and map that into the bins. That is, it does something different.

Reply:
One of the differences to how we did this with `realign` was that with realigned data we required matching coords. For memory/performance reasons we cannot do this with buckets. Also, in the case of realigned data it has led to cumbersome use with unit conversions and bugs (#1305).